### DISCUSSION TITLE
Is there a way to hide the main file?

### DISCUSSION THREAD

**Discussion by petrzavicak** [spatie/laravel-medialibrary#2183] [Q&A]
By default, files are uploaded to the "storage/app/public/media" location. Each photo has its own directory (1,2, etc.) and the image conversion files are located in the conversions subdirectory. However...

I need to make sure that the main uploaded file that the media library works with is not available to website users. In other words, I want to make sure that all the conversions I generate include a watermark, but I don't want to include a watermark in the original file. And of course, I don't want a user who can detect the name of the main file to be able to download the file (without a watermark).

It seems to me to be a very important feature, so I'm surprised that I don't see anyone on the Internet solving this problem.

If the main file is manually moved to another location, subsequent operations - such as file regeneration - https://spatie.be/docs/laravel-medialibrary/v7/converting-images/regenerating-images will not work.

How to achieve? Or did I omit some part of the documentation and is it possible to set the media library to store the main file in the specified location / directory (outside the public folder)?

**Comment by freekmurze** [comment]
You could use a file namer to name your conversions in a way that the name of the original can't be guessed: https://spatie.be/docs/laravel-medialibrary/v9/advanced-usage/naming-generated-files

Another strategy you could use would be to storage the original files in a collection called "originals". You could add a watermark to the original file and then add that file also to a collection named "watermarked" or something similar. 

The `originals` and `watermarked` collection could be put on separate disk. `originals` could even be put in a non-public directory. You could create a controller that let users download the file through PHP, so you can add any authorization logic that you want.

**Reply by petrzavicak** [reply]
Thank you in advance for your answer and I apologize for the longer delay, however, it is only now that I have returned to the topic after a few months.

However - yesterday and today I spent a total of more than 10 hours with the Media Library and I'm slowly running out of nerves. I tried everything, your advices and also a complete rewrite of the Default and Base url generators or the Default Path generator. All unsuccessfully and I always ended up in a dead end.

What's going on:
1. I definitely don't want to create a new collection for "original" files and then place it in a private directory. For a simple reason - I do not want to save the same original file to disk twice (once the original and the same in a special collection). Due to server disk space. This is not the way to go.

2. Overwriting the Path generator so that the name of the original file is difficult to guess is possible, however it is not possible without using a hash in the file name. And I need all the conversions to be saved under the original name as uploaded by the user - that is, if he uploads "My file.jpg", I want users to be able to download "conversions/my-file-conversion-name.jpg" and not "conversions/some-brutal-hash-my-file-conversion-name.jpg ".

So I return to the original question and ask quite simply. In Media library is possible set a different logic to be used to create conversion names. Why isn't it the other way around? WHY AND FOR WHY is it not possible to simply set the original file to use a different name than for conversions? I have a "my-file.jpg" file and want the original file to be saved as "specific-hash-my-file.jpg", but all conversions have been saved as "conversions/my-file-conversion-name.jpg".

After all, it would be enough to add a third method called originalFileName to the conversionFileName and responsiveFileName methods in the DefaultFileNamer class. This is not complicated to implement in the Media library kernel, unfortunately the user has no option to set this behavior in the project. Or I don't know about that ...

I'm asking for advice, I'm already desperate.
